File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(!head) return head;
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* cur = head, *prev = dummy;
        while(cur) {
            while(cur && cur->next && cur->val == cur->next->val) cur = cur->next;
            if(prev->next == cur) {
                prev = prev->next;
            }
            else {
                prev->next = cur;
                prev = cur;
            }
            cur = cur->next;
        }
        return dummy->next;
    }
};

File: 0083-remove-duplicates-from-sorted-list/0083-remove-duplicates-from-sorted-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode cur = head, prev = dummy;
        while(cur != null) {
            while(cur != null && cur.next != null && cur.val == cur.next.val) {
                cur = cur.next;
            }
            if(prev.next == cur) {
                prev = prev.next;
            }
            else {
                prev.next = cur;
                prev = cur;
            }
            cur = cur.next;
        }
        return dummy.next;
    }
}